File: DSA/RATNA_KHANNA_590013141/DAY_76_QUESTION_2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}

bool printAncestors(struct Node* root, int target) {
    if (root == NULL)
        return false;

    // If target node is found, return true
    if (root->data == target)
        return true;

    // Check in left or right subtree
    if (printAncestors(root->left, target) || printAncestors(root->right, target)) {
        printf("%d ", root->data);  // Print the current node
        return true;
    }

    return false;
}

int main() {
    struct Node* root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);

    int target;
    printf("Enter the target node value: ");
    scanf("%d", &target);

    printf("Ancestors of node %d: ", target);
    if (!printAncestors(root, target)) {
        printf("Node not found in the tree.\n");
    }
    printf("\n");

    return 0;
}

File: DSA/VaniAgrawal_590015846/Day_76_Question_2.py
# Print Ancestors of a Node
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def build_tree(values):
    if not values: return None
    nodes = [None if v == "null" else Node(int(v)) for v in values]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

def print_ancestors(root, target):
    if not root: return False
    if root.val == target:
        return True
    if print_ancestors(root.left, target) or print_ancestors(root.right, target):
        print(root.val, end=" ")
        return True
    return False

values = input().split()
target = int(input())
root = build_tree(values)
print_ancestors(root, target)
